prelude.dx

```diff
@@ -233,6 +237,35 @@ randIdx key =
   unif = rand key
   asidx $ floor $ unif * real (size @n)
 
+'min / max etc
+
+minby :: A a::Data. (a->Real)->a->a->a
+minby f x y = select ((f x) < (f y)) x y
+
+min :: Real -> Real -> Real
+min = minby id
+
+minimumby:: A p q::Data. (q->Real)->(p=>q)->q
+minimumby f xs = fold xs.(asidx 0) (for i. minby f xs.i)
```

Comment:
I thought I could do:
```suggestion
minimumby f xs = fold xs.(0@p) (for i. minby f xs.i)
```
to use a index literal.
but that errors complaining about `p` not being defined.

I guess I don';t get how type-parameters in signatures work

Reply:
The syntax for index literals requires the index set to be a literal too, so `0@3` is ok, but `0@p` isn't. I'm surprised it's complaining about `p` not being defined. It should be a parse error.

Note that this case actually does require the unsafe `asidx` cast, because we don't know statically that `xs` is nonempty. An alternative would be to start the fold with an appropriate identity element, like negative infinity, which would give a sensible result even for empty arrays.

Reply:
> An alternative would be to start the fold with an appropriate identity element, like negative infinity, which would give a sensible result even for empty arrays.

We need an appropriate element of type `q`, which I don't think we can get.
e.g. if `q` is `((3=>Int), Real)`
```
>=> minimumby (asNonLin snd) [([1,2,3], -0.2), ([2,3,4], 0.1), ([23,3,4], 3.0)]
([1, 2, 3], -0.2)
```

I'm not even sure it would be fine for `minimum` when you know all is `Real`,
does it make sense to say `minimum [] = Inf` ?
I guess it might.

---

The situation right now is not great for empty arrays, since it does use the unsafe `asidx`
```
>=> xs :: (0=>((3=>Int), Real)) = []

>=> minimumby (asNonLin snd) xs
([140334315231744, -5764607523034234880, -5764607523034234880], -0.0)
```

Maybe this should be added to `bugs`

Reply:
> Maybe this should be added to bugs

Yes, we definitely want a runtime error here rather than reading arbitrary memory :). But I think it's being tracked sufficiently in #31.